plugin/core/protocol.py

```diff
@@ -295,6 +295,13 @@ class SemanticTokenModifiers:
     'items': List[CompletionItem],
 }, total=True)
 
+DocumentLink = TypedDict('DocumentLink', {
+    'range': RangeLsp,
+    'target': DocumentUri,
+    'tooltip': str,
+    'data': Any
+}, total=False)
```

Comment:
Would it be better to have `total=True` and mark `target`, `tooltip` and `data` as optional?
With the way it is now `range` triggers a type error in the part of the code where it's accessed.

Neither `total=False` nor `total=True` is fully accurate. The former because it makes all properties optional while `range` isn't optional, the latter because missing property is not strictly the same as `Optional` property. Thus I would pick whatever is more functional for us.

Reply:
I would think that `total=False` is better, because I directly store the links from the response as `DocumentLink` now, and otherwise the types for links with missing properties wouldn't really be correct as well. This error for `link["range"]` is reported from Pyright, but mypy doesn't seem to have a problem with it. And we already use the same a few lines below for `diagnostic["range"]`, with the same error from Pyright.

Since we know from the specs that "range" is a mandatory property, we could just leave it as it is, or alternatively use the less clear (imo) `link.get("range")`, or add a `# type: ignore` for Pyright.

Reply:
I'll resolve this but I would still prefer the other way around.

There would be an issue with `total=True` had we have to create such typed dict manually because we couldn't create it without providing the optional properties. But since we don't create it manually and rather just use what the LSP server sends us, there wouldn't be an issue.

Proper support for defining such objects seems to be implemented in Python 3.11+ (https://peps.python.org/pep-0655/) and is available in older versions through the https://github.com/python/typing_extensions/ module. We could vendor that module but that's a separate discussion.

Reply:
I mean we could change it, but then we should also change all of the other TypedDict's which currently use `total=False` to `total=True`, e.g. Diagnostic, CompletionItem, etc.
There are countless `reportTypedDictNotRequiredAccess" errors reported from Pyright all over the source code of this repo.

But this could probably be done in another pull request, I guess.